src/rounding.rs

```diff
@@ -0,0 +1,297 @@
+use crate::{
+    options::{TemporalRoundingMode, TemporalUnsignedRoundingMode},
+    TemporalError, TemporalResult,
+};
+
+use std::{
+    cmp::Ordering,
+    ops::{Div, Neg},
+};
+
+use num_traits::{ConstZero, Euclid, FromPrimitive, Signed};
+
+trait Roundable: PartialOrd {
+    fn is_exact(&self) -> bool;
+    fn cmp_rem(&self) -> Option<Ordering>;
+    fn is_cardinal(&self) -> bool;
+    fn result_floor(&self) -> u64;
+    fn result_ceil(&self) -> u64;
+}
+
+pub(crate) trait Round {
+    fn round(&self, mode: TemporalRoundingMode) -> i128;
+    fn round_as_positive(&self, mode: TemporalRoundingMode) -> u64;
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, PartialOrd)]
+pub(crate) struct IncrementRounder<
+    T: Euclid + Div + PartialOrd + Signed + FromPrimitive + ConstZero + Copy,
+> {
+    sign: bool,
+    dividend: T,
+    divisor: T,
+}
+
+impl<T: Euclid + PartialOrd + Signed + FromPrimitive + ConstZero + Copy> IncrementRounder<T> {
+    // ==== PUBLIC ====
+
+    pub(crate) fn from_potentially_negative_parts(number: T, increment: T) -> Self {
+        Self {
+            sign: number / increment > T::ZERO,
+            dividend: number,
+            divisor: increment,
+        }
+    }
+
+    pub(crate) fn from_positive_parts(number: T, increment: T) -> Self {
+        debug_assert!(number / increment > T::ZERO);
+
+        Self {
+            sign: true,
+            dividend: number,
+            divisor: increment,
+        }
+    }
+
+    // ==== PRIVATE ====
+
+    #[inline]
+    fn quotient_abs(&self) -> T {
+        (self.dividend / self.divisor).abs()
+    }
+}
+
+impl Round for IncrementRounder<i128> {
+    fn round(&self, mode: TemporalRoundingMode) -> i128 {
+        let unsigned_rounding_mode = mode.get_unsigned_round_mode(self.sign);
+        let mut rounded = apply_unsigned_rounding_mode(self, unsigned_rounding_mode) as i128;
+        if !self.sign {
+            rounded = rounded.neg();
+        }
+        rounded * self.divisor
+    }
+
+    fn round_as_positive(&self, mode: TemporalRoundingMode) -> u64 {
+        let unsigned_rounding_mode = mode.get_unsigned_round_mode(self.sign);
+        let rounded = apply_unsigned_rounding_mode(self, unsigned_rounding_mode);
+        rounded * self.divisor as u64
+    }
+}
+
+impl Roundable for IncrementRounder<i128> {
+    fn is_exact(&self) -> bool {
+        self.dividend.rem_euclid(self.divisor) == 0
+    }
+
+    fn cmp_rem(&self) -> Option<Ordering> {
+        Some(
+            self.dividend
+                .rem_euclid(self.divisor)
+                .cmp(&self.divisor.div_euclid(2)),
+        )
+    }
```

Comment:
Maybe a better name for the method? `cmp_rem` doesn't really say much about what the method needs to do, which is confusing since it just seems that it is comparing with half the divisor?